content/guides/frameworks/laravel/development-setup.md

```diff
@@ -0,0 +1,342 @@
+# Node.js and asset building are handled in the Nginx container.
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    curl \
+    unzip \
+    libpq-dev \
+    libonig-dev \
+    libssl-dev \
+    libxml2-dev \
+    libcurl4-openssl-dev \
+    libicu-dev \
+    libzip-dev \
+    && docker-php-ext-install -j$(nproc) \
+    pdo_mysql \
+    pdo_pgsql \
+    pgsql \
+    opcache \
+    intl \
+    zip \
+    bcmath \
+    soap \
+    && pecl install redis xdebug \
+    && docker-php-ext-enable redis \
+    && apt-get autoremove -y && apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+# Use ARG to define environment variables passed from the Docker build command or Docker Compose.
+ARG XDEBUG_ENABLED
+ARG XDEBUG_MODE
+ARG XDEBUG_HOST
+ARG XDEBUG_IDE_KEY
+ARG XDEBUG_LOG
+ARG XDEBUG_LOG_LEVEL
+
+# Configure Xdebug if enabled
+RUN if [ "${XDEBUG_ENABLED}" = "true" ]; then \
+    docker-php-ext-enable xdebug && \
+    echo "xdebug.mode=${XDEBUG_MODE}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini && \
+    echo "xdebug.idekey=${XDEBUG_IDE_KEY}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini && \
+    echo "xdebug.log=${XDEBUG_LOG}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini && \
+    echo "xdebug.log_level=${XDEBUG_LOG_LEVEL}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini && \
+    echo "xdebug.client_host=${XDEBUG_HOST}" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini ; \
+    echo "xdebug.start_with_request=yes" >> /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini ; \
+fi
+
+# Set environment variables for user and group ID
+ARG UID=1000
+ARG GID=1000
+
+# Create a new user with the specified UID and GID, reusing an existing group if GID exists
+RUN if getent group ${GID}; then \
+      group_name=$(getent group ${GID} | cut -d: -f1); \
+      useradd -m -u ${UID} -g ${GID} -s /bin/bash www; \
+    else \
+      groupadd -g ${GID} www && \
+      useradd -m -u ${UID} -g www -s /bin/bash www; \
+      group_name=www; \
+    fi
+
+# Dynamically update php-fpm to use the new user and group
+RUN sed -i "s/user = www-data/user = www/g" /usr/local/etc/php-fpm.d/www.conf && \
+    sed -i "s/group = www-data/group = $group_name/g" /usr/local/etc/php-fpm.d/www.conf
+
+
+# Set the working directory
+WORKDIR /var/www
+
+# Copy the entrypoint script
+COPY ./docker/php-fpm/entrypoint.sh /usr/local/bin/entrypoint.sh
+RUN chmod +x /usr/local/bin/entrypoint.sh
+
+# Change the default command to run the entrypoint script
+ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
+
+# Expose port 9000 and start php-fpm server
+EXPOSE 9000
+CMD ["php-fpm"]
+```
+
+This Dockerfile installs the necessary PHP extensions required by Laravel, including database drivers and the Xdebug extension for debugging.
+
+### Create a Dockerfile for Workspace
+
+The workspace container is used to run Artisan commands, Composer, and NPM. Here's the Dockerfile for the workspace:
```

Comment:
In PHP/Laravel, is it a common practice to deploy a sidecar container like this? It seems like a bad practice to have a container that you shell into to perform database migrations or run arbitrary commands in.

If this is a best practice, can you elaborate on what this container is for and why it's idiomatic to do it this way?

Reply:
For a dev environment in Laravel, it’s quite common practice to have a sidecar container. 

Additionally to php cli, which can be reused from the PHP-FPM container, Laravel is using a Node environment for assets building. Additionally, depending on the project, the other various tools can also be included. 

This approach is used in the official [Sail package](https://github.com/laravel/sail/blob/1.x/runtimes/8.4/Dockerfile), as well as in the popular community project [Laradock](https://github.com/laradock/laradock/tree/master/workspace). 

I think we can stick with this approach for dev environment, but I will review the production part. It looks like there the second cli container can be omitted.